### ISSUE TITLE
Route creation order is inconsistent

### ISSUE THREAD

**Issue by JeffGiroux** [F5Networks/f5-declarative-onboarding#147]
### Is your feature request related to a problem? Please describe.
Google GCP has unique routing, and this requires the VMs to deploy with additional routes in order for traffic to work. DO has a Route class that supports a target of 'gw', but this only supports IP addresses. If you add a vlan interface like 'external', the declaration is bad.

These are my routes...
```
        "ext_gw_interface": {
            "class": "Route",
            "gw": "external",
            "network": "10.1.10.1/32",
            "mtu": 1460
        },
        "ext_rt": {
            "class": "Route",
            "gw": "10.1.10.1",
            "network": "10.1.10.0/24",
            "mtu": 1460
        },
        "default": {
            "class": "Route",
            "gw": "10.1.10.1",
            "network": "default",
            "mtu": 1460
        }
```

Error...
```
Wed, 10 Jun 2020 21:41:43 GMT - finest: [f5-declarative-onboarding: restWorker.js] Got onboarding request.
Wed, 10 Jun 2020 21:41:43 GMT - info: [f5-declarative-onboarding: restWorker.js] Bad declaration: [{"keyword":"format","dataPath":".declaration.Common['ext_gw_interface'].gw","schemaPath":"network.schema.json#/allOf/4/then/properties/gw/anyOf/0/format","params":{"format":"ipv4"},"message":"should match format \"ipv4\""},{"keyword":"format","dataPath":".declaration.Common['ext_gw_interface'].gw","schemaPath":"network.schema.json#/allOf/4/then/properties/gw/anyOf/1/format","params":{"format":"ipv6"},"message":"should match format \"ipv6\""},{"keyword":"anyOf","dataPath":".declaration.Common['ext_gw_interface'].gw","schemaPath":"network.schema.json#/allOf/4/then/properties/gw/anyOf","params":{},"message":"should match some schema in anyOf"}]
```

### Describe the solution you'd like
Add 'interface' as a parameter to the Route class.
https://clouddocs.f5.com/products/extensions/f5-declarative-onboarding/latest/schema-reference.html#route

This will allow me to choose gw or interface in my declaration.

Example with 'interface'...
```
        "ext_gw_interface": {
            "class": "Route",
            "interface": "external",
            "network": "10.1.10.1/32",
            "mtu": 1460
        },
```

### Describe alternatives you've considered
I have to onboard my BIG-IP with tmsh commands right now as a workaround.
```
tmsh create sys management-ip ${MGMTADDRESS}/32
tmsh create sys management-route mgmt_gw network ${MGMTGATEWAY}/32 type interface mtu 1460
tmsh create sys management-route mgmt_net network ${MGMTNETWORK}/${MGMTMASK} gateway ${MGMTGATEWAY} mtu 1460
tmsh create sys management-route default gateway ${MGMTGATEWAY} mtu 1460
tmsh create net vlan external interfaces add { 1.0 } mtu 1460
tmsh create net self self_external address ${INT2ADDRESS}/32 vlan external
tmsh create net route ext_gw_interface network ${INT2GATEWAY}/32 interface external
tmsh create net route ext_rt network ${INT2NETWORK}/${INT2MASK} gw ${INT2GATEWAY}
tmsh create net route default gw ${INT2GATEWAY}
```

**Comment by dstokesf5** [comment]
Thank you for your feedback. I have added this to our internal product backlog as AUTOTOOL-1661.

**Comment by RobCupples** [comment]
@JeffGiroux 

If you want to use a VLAN interface I believe you need to specify it with "target" instead of "gw".

**Comment by JeffGiroux** [comment]
I'm pretty sure I tried that originally before opening ticket and it failed. Per API docs...

Target = "The VLAN or Tunnel for the Route"

However, I need 'interface'. Based on API docs, target only supports vlan and tunnel. I need target to support, vlan, tunnel, AND interface.

I'll try to test again, but I'm pretty sure it won't work. Will update when I get a chance to test. Here's what I plan to test...snippet...

Notice 'target=external' instead of 'gw=external'. Will try this...
```
        "ext_gw_interface": {
            "class": "Route",
            "target": "external",
            "network": "10.1.10.1/32",
            "mtu": 1460
        },
        "ext_rt": {
            "class": "Route",
            "gw": "10.1.10.1",
            "network": "10.1.10.0/24",
            "mtu": 1460
        },
        "default": {
            "class": "Route",
            "gw": "10.1.10.1",
            "network": "default",
            "mtu": 1460
        }

**Comment by JeffGiroux** [comment]
Hmm...at first attempt with Postman. Fail...

```
                "params": {
                    "additionalProperty": "target"
                },
                "message": "should NOT have additional properties"
```

**Comment by RobCupples** [comment]
What tmsh calls "interface" declarative onboarding calls "target".  We will try to reproduce.

**Comment by JeffGiroux** [comment]
Listed in ticket opening :) Repasting...

```
tmsh create sys management-ip ${MGMTADDRESS}/32
tmsh create sys management-route mgmt_gw network ${MGMTGATEWAY}/32 type interface mtu 1460
tmsh create sys management-route mgmt_net network ${MGMTNETWORK}/${MGMTMASK} gateway ${MGMTGATEWAY} mtu 1460
tmsh create sys management-route default gateway ${MGMTGATEWAY} mtu 1460
tmsh create net vlan external interfaces add { 1.0 } mtu 1460
tmsh create net self self_external address ${INT2ADDRESS}/32 vlan external
tmsh create net route ext_gw_interface network ${INT2GATEWAY}/32 interface external
tmsh create net route ext_rt network ${INT2NETWORK}/${INT2MASK} gw ${INT2GATEWAY}
tmsh create net route default gw ${INT2GATEWAY}
```

**Comment by RobCupples** [comment]
I haven't been able to reproduce this problem.  Here is a snippet trimmed out of the declaration I used.

```
        "Common": {
            "class": "Tenant",
            "external": {
                "class": "VLAN",
                "mtu": 1500,
                "interfaces": [
                    {
                        "name": "1.1",
                        "tagged": false
                    }
                ]
            },
            "external-self": {
                "class": "SelfIp",
                "address": "10.20.0.100/24",
                "vlan": "external",
                "allowService": "none",
                "trafficGroup": "traffic-group-local-only"
            },
            "ext_gw_interface": {
                "class": "Route",
                "target": "external",
                "network": "10.1.10.1/32",
                "mtu": 1460
            },
            "ext_rt": {
                "class": "Route",
                "gw": "10.1.10.1",
                "network": "10.1.10.0/24",
                "mtu": 1460
            },
            "default": {
                "class": "Route",
                "gw": "10.1.10.1",
                "network": "default",
                "mtu": 1460
            }
        }
```

**Comment by JeffGiroux** [comment]
OK, latest testing has gotten me closer to a more exact error. Flow of events...

1. Started with DO 1.13.0 and BIG-IP 15.1.0.4
2. Opened ticket a few months ago thinking 'interface' needed to support 'target'
3. Tested my declaration with external routes in declaration but received errors when testing with 'interface' target as noted in ticket opening
4. Worked with Rob C, added self IP and vlan to declaration along with route for 'external' network
5. Tested again and same error
```
        "result": {
            "class": "Result",
            "code": 400,
            "status": "ERROR",
            "message": "bad declaration",
            "errors": [
                {
                    "keyword": "additionalProperties",
                    "dataPath": ".declaration.Common['ext_gw_interface']",
                    "schemaPath": "network.schema.json#/allOf/4/then/additionalProperties",
                    "params": {
                        "additionalProperty": "target"
                    },
                    "message": "should NOT have additional properties"
                }
            ]
        },
```
6. Upgraded DO to 1.14.0, BIG-IP still on 15.1.0.4, same exact declaration
7. Tested new VE instance again...this time new error. Keep in mind that tmsh is still creating internal vlan/selfip/route before DO. I only moved external nic stuff to DO. In this error, DO is complaining about the inability to change internal. Understandable as it is trying to wipe all and control the L2-L3 objects. Also, notice the /32 "not directly connected" error. More on that later...
```
        "message": "rollback failed",
        "errors": [
            "01070265:3: The VLAN (/Common/internal) cannot be deleted because it is in use by a self IP (/Common/self_internal).",
            "01070330:3: Static route gateway 10.1.20.1 is not directly connected via an interface."
        ],
        "result": {
            "class": "Result",
            "code": 500,
            "status": "ERROR",
            "message": "rollback failed",
            "errors": [
                "01070265:3: The VLAN (/Common/internal) cannot be deleted because it is in use by a self IP (/Common/self_internal).",
                "01070330:3: Static route gateway 10.1.20.1 is not directly connected via an interface."
            ]
        },
```
8. I then updated my repo, removed all tmsh commands for internal NIC stuff so DO would do internal nic
9. Now DO declaration has external self IP/vlan/route and internal self IP/vlan/route
10. I test a new VE deployment and new error...
```
        "status": "ERROR",
        "message": "rollback failed",
        "errors": [
            "01070330:3: Static route gateway 10.1.10.1 is not directly connected via an interface.",
            "010719fd:3: No IPv4 self IP exists on VLAN (/Common/external) for static route (/Common/ext_gw_interface)"
        ],
```

In GCP, routing is different. You have to target interfaces and set self IPs at /32. This means…the gateway address of 10.1.10.1 for example is truly unreachable to a VE with a self IP in 10.1.10.100/32 for example. Because…10.1.10.100/32 does not have a directly connected interface to 10.1.10.1/32. Different networks…/32. Thx GCP ☹

As a result, tmsh commands take this fine on the BIG-IP. No error. I originally opened this ticket thinking that DO didn't support 'interface' as a target. But...I don't think the issue is with DO and lack of target 'interface'. That actually works. The real problem in this ticket is DO's lack of ability to add a route similar to tmsh in situations where the self IP is on a /32...for cloud requirements like in GCP.

Allow DO to add routes in the way I expressed above (TMSH lets you do it), and you will have the fix!

Full declaration error output...
```
[
    {
        "id": "b3be810e-03ba-4b6b-8cb2-239fa50f103c",
        "selfLink": "https://localhost/mgmt/shared/declarative-onboarding/task/b3be810e-03ba-4b6b-8cb2-239fa50f103c",
        "code": 500,
        "status": "ERROR",
        "message": "rollback failed",
        "errors": [
            "01070330:3: Static route gateway 10.1.10.1 is not directly connected via an interface.",
            "010719fd:3: No IPv4 self IP exists on VLAN (/Common/external) for static route (/Common/ext_gw_interface)"
        ],
        "result": {
            "class": "Result",
            "code": 500,
            "status": "ERROR",
            "message": "rollback failed",
            "errors": [
                "01070330:3: Static route gateway 10.1.10.1 is not directly connected via an interface.",
                "010719fd:3: No IPv4 self IP exists on VLAN (/Common/external) for static route (/Common/ext_gw_interface)"
            ]
        },
        "declaration": {
            "schemaVersion": "1.0.0",
            "class": "Device",
            "async": true,
            "label": "Onboard BIG-IP standalone 2nic",
            "Common": {
                "class": "Tenant",
                "hostname": "giroux4-f5vm01.c.f5-4136-mspteam-dev.internal",
                "dbVars": {
                    "class": "DbVariables",
                    "ui.advisory.enabled": true,
                    "ui.advisory.color": "green",
                    "ui.advisory.text": "/Common/hostname",
                    "config.allow.rfc3927": "enable"
                },
                "myDns": {
                    "class": "DNS",
                    "nameServers": [
                        "169.254.169.254",
                        "8.8.8.8",
                        "2001:4860:4860::8844"
                    ],
                    "search": [
                        "c.f5-4136-mspteam-dev.internal",
                        "google.internal",
                        "f5.com"
                    ]
                },
                "myNtp": {
                    "class": "NTP",
                    "servers": [
                        "0.us.pool.ntp.org",
                        "1.pool.ntp.org",
                        "2.pool.ntp.org"
                    ],
                    "timezone": "UTC"
                },
                "myProvisioning": {
                    "class": "Provision",
                    "ltm": "nominal"
                },
                "external": {
                    "class": "VLAN",
                    "mtu": 1460,
                    "interfaces": [
                        {
                            "name": "1.0",
                            "tagged": false
                        }
                    ],
                    "cmpHash": "default",
                    "failsafeEnabled": false,
                    "failsafeAction": "failover-restart-tm",
                    "failsafeTimeout": 90
                },
                "external-self": {
                    "class": "SelfIp",
                    "address": "10.1.10.80/32",
                    "vlan": "external",
                    "allowService": "none",
                    "trafficGroup": "traffic-group-local-only"
                },
                "ext_gw_interface": {
                    "class": "Route",
                    "target": "external",
                    "network": "10.1.10.1/32",
                    "mtu": 1460
                },
                "ext_rt": {
                    "class": "Route",
                    "gw": "10.1.10.1",
                    "network": "10.1.10.0/24",
                    "mtu": 1460
                },
                "default": {
                    "class": "Route",
                    "gw": "10.1.10.1",
                    "network": "default",
                    "mtu": 1460
                },
                "internal": {
                    "class": "VLAN",
                    "mtu": 1460,
                    "interfaces": [
                        {
                            "name": "1.2",
                            "tagged": false
                        }
                    ],
                    "cmpHash": "default",
                    "failsafeEnabled": false,
                    "failsafeAction": "failover-restart-tm",
                    "failsafeTimeout": 90
                },
                "internal-self": {
                    "class": "SelfIp",
                    "address": "10.1.20.31/32",
                    "vlan": "internal",
                    "allowService": "default",
                    "trafficGroup": "traffic-group-local-only"
                },
                "int_gw_interface": {
                    "class": "Route",
                    "target": "internal",
                    "network": "10.1.20.1/32",
                    "mtu": 1460
                },
                "int_rt": {
                    "class": "Route",
                    "gw": "10.1.20.1",
                    "network": "10.1.20.0/24",
                    "mtu": 1460
                }
            }
        },
        "httpStatus": 200
    }
]
```

**Comment by RobCupples** [comment]
Ok.  If I use the entire block of VLANs, SelfIps, and Routes you show above I am having similar issues.  If I build up my declaration one item at a time posting each time I add an item it works.  If I then remove them all at once and post again it is having problems deleting the items so there appears to be an ordering issue going on.
  
CC: @dstokesf5

**Comment by dstokesf5** [comment]
I have confirmed that this is an ordering issue. Routes are not created in a consistent order, and this declaration requires some routes be created before others. By building up the declaration a piece at a time, the correct order can be guaranteed. DO will need to sort creation of routes, or at least consistently create them in the order that they appear in the declaration. I have created AUTOTOOL-1845 for this work.

Note AUTOTOOL-1661 was closed since the original request was invalid.
